sound/soc/sof/virtio-fe.c

```diff
@@ -19,112 +20,418 @@
  * The virtIO message Q will use the *exact* same IPC structures as we currently
  * use in the mailbox.
  *
- * Guest OS SOF core -> SOF FE -> virtIO Q -> SOF BE ->
- * System OS SOF core -> DSP
- *
  * The mailbox IO and TX/RX msg functions below will do IO on the virt IO Q.
  */
 
-#include <linux/delay.h>
-#include <linux/fs.h>
-#include <linux/slab.h>
 #include <linux/device.h>
+#include <linux/firmware.h>
 #include <linux/interrupt.h>
 #include <linux/module.h>
-#include <linux/dma-mapping.h>
-#include <linux/firmware.h>
-#include <linux/device.h>
+#include <linux/scatterlist.h>
 #include <linux/virtio.h>
+#include <linux/virtio_ring.h>
 #include <sound/sof.h>
 #include <uapi/sound/sof-fw.h>
+#include <uapi/sound/sof-ipc.h>
 
-#include "sof-priv.h"
 #include "ops.h"
-#include "intel.h"
+#include "sof-priv.h"
+#include "intel/hda.h"
+
+static const char *const sof_vq_names[SOF_VIRTIO_NUM_OF_VQS] = {
+	SOF_VIRTIO_IPC_CMD_TX_VQ_NAME,
+	SOF_VIRTIO_IPC_CMD_RX_VQ_NAME,
+	SOF_VIRTIO_IPC_NOT_TX_VQ_NAME,
+	SOF_VIRTIO_IPC_NOT_RX_VQ_NAME,
+};
+
+struct sof_virtio_priv {
+	struct snd_sof_pdata *sof_pdata;
+	struct platform_device *pdev_pcm;
+};
+
+static const struct sof_dev_desc virt_desc = {
+	.nocodec_fw_filename = NULL,
+	.nocodec_tplg_filename = "intel/sof-apl-nocodec.tplg",
+	.resindex_lpe_base	= -1,
+	.resindex_pcicfg_base	= -1,
+	.resindex_imr_base	= -1,
+	.irqindex_host_ipc	= -1,
+	.resindex_dma_base	= -1,
+};
+
 
 /*
  * IPC Firmware ready.
  */
-static int virtio_fe_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
+static int vfe_is_ready(struct snd_sof_dev *sdev)
+{
+	/* is message still pending */
+	if (sdev->vfe->msg)
+		return 0;
+
+	/* ready for next message */
+	return 1;
+}
+
+static int vfe_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
 {
-	/* not needed for FE ? */
 	return 0;
+};
+
+/* used to send IPC to BE */
+static int vfe_send_msg(struct snd_sof_dev *sdev,
+			struct snd_sof_ipc_msg *msg)
+{
+	struct sof_vfe *vfe;
+	struct scatterlist sgs[2];
+	int ret = 0;
+
+	vfe = sdev->vfe;
+
+	sg_init_table(sgs, 2);
```

Comment:
Can we use macros for these magic numbers, I've seen them used in other places for arrays too.